pkg/codegen/schema/schema.go

```diff
@@ -925,63 +941,150 @@ func (t *types) bindPrimitiveType(name string) (Type, error) {
-		case "pulumi.json#/Any":
-			return AnyType, nil
+// externalSchemaRef contains the parsed fields for an external schema ref.
+type externalSchemaRef struct {
+	Package string          // The package component of the schema ref
+	Version *semver.Version // The version component of the schema ref
+}
+
+// typeSpecRef contains the parsed fields from a type spec reference.
+type typeSpecRef struct {
+	*externalSchemaRef
+
+	URL   *url.URL // The parsed URL
+	Token string   // The type token
+	Kind  string   // The kind of reference, either 'resources' or 'types'
+}
+
+// Regexs used to parse type spec refs. These are declared at the package scope to avoid repeatedly recompiling them.
+var refPathRegex = regexp.MustCompile(`^/(?P<package>\w+)/v?(?P<version>\d+\.\d+\.\d+)/schema\.json$`)
+var refFragmentRegex = regexp.MustCompile(`^/(?P<kind>resources|types)/(?P<token>.*:.*:.*)$`)
+
+func parseTypeSpecRef(ref string) typeSpecRef {
+	parsedURL, err := url.Parse(ref)
+	contract.AssertNoErrorf(err, "failed to parse ref URL: %s", ref)
+
+	fragment, err := url.PathUnescape(parsedURL.Fragment)
+	contract.AssertNoError(err)
+
+	fragmentMatch := refFragmentRegex.FindStringSubmatch(fragment)
+	contract.Assertf(len(fragmentMatch) == 3, "failed to parse fragment: %s", fragment)
+
+	result := typeSpecRef{
+		URL:   parsedURL,
+		Kind:  fragmentMatch[1],
+		Token: fragmentMatch[2],
+	}
+
+	if len(parsedURL.Path) > 0 {
+		pathMatch := refPathRegex.FindStringSubmatch(parsedURL.Path)
+		contract.Assert(len(pathMatch) == 3)
+
+		pkg, versionToken := pathMatch[1], pathMatch[2]
+		version, err := semver.Parse(versionToken)
+		contract.AssertNoErrorf(err, "failed to parse package version: %s", versionToken)
+
+		result.externalSchemaRef = &externalSchemaRef{
+			Package: pkg,
+			Version: &version,
 		}
+	}
 
-		// Parse the ref and look up the type in the type map.
-		switch {
-		case strings.HasPrefix(spec.Ref, "#/types/"):
-			token, err := url.PathUnescape(spec.Ref[len("#/types/"):])
-			if err != nil {
-				return nil, err
-			}
+	return result
+}
 
-			if typ, ok := t.objects[token]; ok {
-				return typ, nil
-			}
-			if typ, ok := t.enums[token]; ok {
-				return typ, nil
-			}
-			typ, ok := t.tokens[token]
+func (t *types) bindTypeSpecRef(spec TypeSpec) (Type, error) {
+	// Explicitly handle built-in types so that we don't have to handle this type of path during ref parsing.
+	switch spec.Ref {
+	case "pulumi.json#/Archive":
+		return ArchiveType, nil
+	case "pulumi.json#/Asset":
+		return AssetType, nil
+	case "pulumi.json#/Json":
+		return JSONType, nil
+	case "pulumi.json#/Any":
+		return AnyType, nil
+	}
+
+	ref := parseTypeSpecRef(spec.Ref)
+
+	if ref.externalSchemaRef != nil {
+		pkg, err := t.loader.LoadPackage(ref.externalSchemaRef.Package, ref.externalSchemaRef.Version)
+		if err != nil {
+			return nil, fmt.Errorf("resolving package %v: %w", ref.URL, err)
+		}
+
+		switch ref.Kind {
+		case "types":
+			typ, ok := t.named[ref.Token]
```

Comment:
As discussed, we should just add a `GetType` function to `Package` and use that instead of switching on `Kind`. That will ensure that any packages linked together using the same loader will have properly interned types, since each package is responsible for the types it defines.

Reply:
Ah, I did part of that conversion, but forgot that `.named` includes both resources and types. Will update.